src/cpp/session/modules/SessionProjectTemplate.cpp

```diff
@@ -0,0 +1,384 @@
+      boost::sregex_token_iterator end;
+      
+      for (; it != end; ++it) {
+         // invoke parser on current record
+         std::map<std::string, std::string> fields;
+         std::string errorMessage;
+         error = text::parseDcfFile(*it, true, &fields, &errorMessage);
+         if (error)
+            continue;
+         
+         // populate project template description based on fields
+         error = ProjectTemplateDescription::populate(resourcePath, fields, pDescription);
+         if (error)
+            continue;
+      }
+      
+      return error;
+   }
+   
+private:
+   void onIndexingStarted()
+   {
+      pRegistry_ = boost::make_shared<ProjectTemplateRegistry>();
+   }
+
+   void onWork(const std::string& pkgName, const core::FilePath& resourcePath)
+   {
+      Error error;
+      
+      // loop over discovered files and attempt to read template descriptions
+      std::vector<FilePath> children;
+      error = resourcePath.children(&children);
+      if (error)
+         LOG_ERROR(error);
+      
+      BOOST_FOREACH(const FilePath& childPath, children)
+      {
+         // skip files that don't have a dcf extension
+         if (childPath.extension() != ".dcf")
+            continue;
+         
+         ProjectTemplateDescription description;
+         Error error = parseResourceFile(pkgName, childPath, &description);
+         if (error)
+         {
+            LOG_ERROR(error);
+            continue;
+         }
+         
+         normalize(&description);
+         
+         pRegistry_->add(pkgName, description);
+      }
+   }
+
+   void onIndexingCompleted()
+   {
+      // index a project template file in the RStudio options folder
+      FilePath localTemplatesPath =
+            module_context::resolveAliasedPath("~/.R/" kRStudioProjectTemplatesPath);
+      
+      if (localTemplatesPath.exists())
+         onWork(kProjectTemplateLocal, localTemplatesPath);
+      
+      // update global registry
+      projectTemplateRegistry() = *pRegistry_;
+      
+      // add known project templates
+      addKnownProjectTemplates();
+      
+      // execute any callbacks waiting for indexing to complete
+      executeCallbacks();
+      
+      // notify client
+      notifyClient();
+   }
+   
+private:
+   Error validateFields(std::map<std::string, std::string>& fields,
+                        const ErrorLocation& location)
+   {
+      std::vector<std::string> missingFields;
+      const char* expected[] = {"binding", "title", "subtitle", "caption"};
+      for (std::size_t i = 0; i < sizeof(expected); ++i)
+      {
+         const char* field = expected[i];
+         if (!fields.count(field) || fields[field].empty())
+            missingFields.push_back(field);
+      }
+      
+      if (missingFields.empty())
+         return Success();
+      
+      std::string reason =
+            "invalid project template description: missing or empty fields "
```

Comment:
Could we also log the location of the file in which these fields were expected? You can just add this as a property on the error; otherwise I think the message in the logs will be hard to associate with a particular package.